### ISSUE TITLE
How to calculate the success rate and the problem of parameters.

### ISSUE THREAD

**Issue by chenzixuan99** [sufengniu/GVIN#2]
Hi! After reading your paper and your code,I have some questions:
1.In the code about reinforcement learning,the success rate can be calculated when the statebatchsize =1 , because the the number of state to be predicted is 1, however in the imitation learning, because the statebatchsize is not 1, then a graph contains multiple states that need to be predicted synchronously.May I ask,how to calculate the success rate in the case where the number of states to be predicted is not 1?

2.According to your paper, some of the parameters are inconsistent with the code, such as RMSProp decay factor (0.999), initial value of exploration rate (0.3). Do these differences cause my results to fail to reach your experimental results? Are these parameters set according to the paper or the code?

Looking forward to your reply! Thank you!

**Comment by sufengniu** [comment]
Hi albzni,

Thank you for the feedback
1. you can change the statebatchsize more than 1 for testing, the success rate is defined as given a starting state, whether the agent could reach the goal state. We haven't try statebatchsize > 1 for testing cases, our experiment use one trial for each map. If you want to more statebatchsize for each testing map, what pops out my mind is you may need to tweak the code that using an external for loop to iterate different starting point instead using batch, since the trajectory length might be different. 

2. The code shown RMSProp decay factor (0.999), initial value of exploration rate (0.3) is not the one we used in the paper. We haven't got time to clean the code, so that the code is very messy (Sorry about that), and some parameters are not correctly set. what we observe is that the exploration rate plays important roles, the exploration rate starting from 1 decaying to 0. Can you elaborate more on your failed experiments?

**Comment by chenzixuan99** [comment]
Hi,sufengniu.Thank you for your reply!

I ran irregular/IL/train.py and /irregular/RL/rl_nstep.py, and the results were Testing Accuracy: 89.29% and Current_Succ_Rate: 99.55%, Avg Difference: 0.071, Avg Reward: 0.969. These experimental results are quite different from the results shown in your paper, so I would like to ask: 1) How do you calculate the success rate, path difference and expected reward under the IL training algorithm shown in your paper? 2) If the training is performed with data of 100 nodes, is the parameter in the network changed? For example, "ch_q".

Looking forward to your reply!

**Comment by sufengniu** [comment]
Hello Albzni,

Sorry for very late reply, since I was extremely busy recently. I think the result you got is very similar to the original paper Table 2 embedding based kernel. Can you point out which part you think is quite anomaly? 

Also, we found that this code is too mess, and part of code is re-usable and not reproduce-able. We are planning to re-write the code an integrate with some our current on-going work together. Please stay tuned.

Sufeng

**Comment by chenzixuan99** [comment]
Hi,sufengniu. Thank you for your reply!

I am very glad to hear that you are planning to re-write the code! Looking forward to your on-going work!

Is it convenient for you to provide the code of generating the graph? Because I want to do some tests on different graphs, such as 5 nodes, 20 nodes, etc. If you can provide it, this will help me a lot !
Thank your so much!

**Comment by eccentricpzy** [comment]
Thank you for providing GVIN code to make it easy to reproduce.
1. I take the success rate code in rl_nstep.py and port it to /irregular/IL folder, and most of experiments achieve similar results as the paper, but when we trained on 10 nodes and tested on 100 nodes, the success rate was 64.64%, much higher than the results in the paper (6.73%). 
2. At the same time, I want to verify a problem: in IL, when the training set changes from 10 nodes to 100 nodes, the accuracy of the test set of 100 nodes has not been significantly improved (56.14%->58.90 %). If we provide more nodes during training or reduce the number of nodes during testing, can GVIN improve the prediction accuracy in the test set?
3. If it is convenient, could you provide a mat file that generates the data and the file to calculate the successful rate? So that I can verify the above problem by myself.

**Comment by sufengniu** [comment]
Hello @eccentricpzy  sorry for the late reply, recently, I was pretty busy.

1. ok, that is interesting, I need to check it whether where the mistake is. I did remembered that when using imitation learning, we encounter heavily overfitting. I need to double check on it

2. yes, we also observe this issue as well (the prediction accuracy is not improved when your training image is large). Here is the intuition: whether you train on the small graph or large graph, the graph embedding are all normalized between 0 to 1, and thereby the performance is actually not depends on the graph scale. To improve the prediction accuracy, I think two things might worth to look at: 1) GVIN is special case of graph neural network, which edge updates in GNN is kernel function in GVIN, we found that in many scenarios the P matrix sometime is not meaningful. you might want to see if using recurrent network to learn the P. 2) is the policy, unlike VIN, use max Q as policy, our last step use max V since the number of action keeps changing on different position, we haven't dig deeper into it, but I think this might be a problem in the irregular graph

3. the data generation is from another author, who is graduated, I can try to pin him see if he could share the code.

**Comment by sufengniu** [comment]
Hi @albzni 

Sorry for the very late reply. the generation code actually is from my colleague mates, let me ask if he would share it, I will let you know and update it to the github, I think he used Matlab to generate it

Thank you

**Comment by Lunatickana** [comment]
hi! thank you for upload your code. it's very useful for my research.

and i want to know can you upload the real world data(just like the New York datafile which in the paper)
it will help a lot, thank you.
best wish

**Comment by chenzixuan99** [comment]
Hi,sufengniu. Thank you for your reply! I'm very sorry for the late reply.

Thank you very much for helping me to ask your colleague mates! Hoping that he can share the generation code because it can help me a lot.  Looking forward to your reply!

Best wishes.

**Comment by eccentricpzy** [comment]
Thanks for your replay. I’m sorry for the late reply because I still am trying on the 3rd mission on my own.
1.	We also noticed the over-fitting problem: In IL setting, only a small amount of training data is needed to achieve considerable results, but as the training data increases, the performance decreases. Sadly, now we have not yet thought of an idea to solve it.

2.	1) In the embedding-based kernel, we think the matrix P is just one of many flexible models that can get the optimal strategy because it can be adjusted through the neural network, as VIN does. I have some doubts about the other part of the embedding-based kernel —— a fixed symmetric matrix based on the adjacency matrix. Does this matrix mainly consider the degree of each node? 

     2) In irregular graph, we also think the maxV is more accurate than maxQ because we didn’t need to consider actions, so we will keep your idea in your paper ^-^.

3.	What troubles me now is still unable to generate mat file. I tried a lot but failed. At present, there are few experiments that can be done with two mat files, therefore it’s very hard for me to expand your work. So, when you have time, I still hope that you can provide some files with other number of nodes or data generation.

**Comment by sufengniu** [comment]
Hello @eccentricpzy sorry for the late reply,

1. thank you for pointing out, we also observed this problems, and this is a problem we also try to solve in the near future

2. i) you are right, it is normalized based on the node degree.

3. I have asked my collaborator, we missed that code. The code is based on random geometric graph using matlab, nothing fancy. Sorry about that.
